src/transformers/models/mega/modeling_mega.py

```diff
@@ -0,0 +1,2311 @@
+
+    def step(self, inputs, length, past_state=None):
+        if length == 1:
+            return self.one_step(inputs, past_state=past_state)
+
+        # (kernel_dim X ema_projection_size X 1)
+        p_coeff, q_coeff = self.coeffs()
+        # (kernel_dim X ema_projection_size X 1+sequence_length)
+        vander = torch.arange(length + 1).to(p_coeff).view(1, 1, length + 1) * torch.log(q_coeff)
+        vander = torch.exp(vander)
+        if past_state is not None:
+            # (kernel_dim X ema_projection_size X sequence_length) * (kernel_dim X ema_projection_size X 1)
+            # -> (kernel_dim X ema_projection_size X sequence_length)
+            past_ema_proj = vander[:, :, 1:] * (self.g_param * self.scale).unsqueeze(-1)
+            # past_state will be (batch_size, kernel_dim, ema_projection_size)
+            past_ema_state = torch.einsum("bdn,dnl->bdl", past_state, past_ema_proj)
+            # (kernel_dim X ema_projection_size) * (batch_size X kernel_dim X ema_projection_size)
+            # -> (batch_size X kernel_dim X ema_projection_size)
+            past_vandermonde = vander[:, :, -1] * past_state
+        else:
+            past_ema_state = None
+            past_vandermonde = None
+
+        # (kernel_dim X ema_projection_size X sequence_length)
+        vander = vander[:, :, :-1]
+        kernel = (p_coeff * self.b_param) * vander
+        kernel_proj = torch.einsum("dnl,dn->dl", kernel, self.g_param * self.scale)
+
+        kernel_fourier = torch.fft.rfft(kernel_proj.float(), n=2 * length)
+        inputs_fourier = torch.fft.rfft(inputs.float(), n=2 * length)
+        # (batch_size X kernel_dim X sequence_length)
+        out = torch.fft.irfft(inputs_fourier * kernel_fourier, n=2 * length)[..., 0:length]
+        out = out.type_as(inputs)
+        if past_ema_state is not None:
+            out = out + past_ema_state
+
+        updated_hidden_state = torch.einsum("bdl,dnl->bdn", inputs, torch.flip(kernel, dims=[2]))
+        if past_vandermonde is not None:
+            updated_hidden_state = updated_hidden_state + past_vandermonde
+        # return a tuple:
+        # (sequence_length, batch_size, kernel_dim)
+        # (batch_size, kernel_dim, ema_projection_size)
+        return out.permute(2, 0, 1), updated_hidden_state
+
+    def one_step(self, inputs, past_state=None):
+        p_coeff, q_coeff = self.coeffs()
+        # (kernel_dim X ema_projection_size) x (batch_size X kernel_dim X 1)
+        # -> (batch_size X kernel_dim X ema_projection_size)
+        updated_state = (p_coeff * self.b_param).squeeze(-1) * inputs
+        if past_state is not None:
+            updated_state = updated_state + q_coeff.squeeze(-1) * past_state
+        # (batch_size X kernel_dim)
+        out = torch.einsum("bdn,dn->bd", updated_state, self.g_param * self.scale)
+        # (1 X batch_size X kernel_dim), (batch_size X kernel_dim X ema_projection_size)
+        return out.unsqueeze(0), updated_state
+
+    def forward(
```

Comment:
Let's split this into sub layers. We can have `MegaDampedEMALayer`, then the `MegaGatedAttentionUnit` and `MegaSingleHeadAttentionLayer`. This will make the code more understanble w.r.t the paper IMO. Then the `MegaLayer` is the final wrapper.

Reply:
Also there is no self attention in this layer (the doc says so)

Reply:
Docstring updated in upcoming commit - after looking back at this, I think this `MultiDimensionDampedEMA` class is fairly self-contained as it's only performing the EMA component, and it aligns with the paper's EMA sub-layer.